src/main/scala/Ch3.scala

```diff
@@ -125,3 +125,65 @@ object Ch3Exercises {
     g: A => C contradicts with A => B, therefore not well typed
     */
 }
+
+object Ch3Ex2 {
+
+  // Problem 1
+  sealed trait CellState {
+    def isBomb: Boolean = this match {
+      case BombCell() => true
+      case _ => false
+    }
+  }
+
+  final case class ClosedCell() extends CellState
+  final case class BombCell() extends CellState
+  final case class OpenCell(neighborBombs: Int) extends CellState
+
+  // Problem 2
+  def numCellsShowingZeroNeighborBombs(cells: Seq[Seq[CellState]]): Int = {
+    cells.flatten.filter{
+      _ match {
+        case OpenCell(0) => true
+        case _ => false
+      }
+    }.length
+  }
+
+  // Problem 3
+  sealed trait RootOfLinear
+  final case class NoRoot() extends RootOfLinear
+  final case class OneRoot(x: Double) extends RootOfLinear
+  final case class AllXRoots() extends RootOfLinear
+  def solve1(a: Double, b: Double): RootOfLinear = (a, b) match {
+    case (0, 0) => AllXRoots()
+    case (0, _) => NoRoot()
+    case (a, b) => OneRoot(-b / a)
+  }
+
+  // Problem 4
+  def solve1(pairs: Seq[(Double, Double)]): Seq[Double] = {
+    pairs.map{ case (a, b) => solve1(a, b) }.map {
```

Comment:
Use .flatMap instead of .map().flatten

Reply:
Oh wow that's a really good catch. Thanks!